### PR TITLE
Support a globally-installable escript

### PR DESCRIPTION
**Pull request by gworkman**
## tl;dr

I added support to install NervesHubCLI as an escript, to evaluate whether this provides a better experience for working with NervesHub from the command line without configuring it with every project. It's quite rough around the edges

## Problem

I recently was invited to try out the NervesCloud beta by Lars and Josh. This was my first time setting up NervesHub, and unfortunately I found the process to be confusing. While NervesHubLink makes sense to add as a project dependency, I couldn't quite figure out how to work with the mix tasks for the CLI. 

I know that 2.0 is under active development and the docs are getting better with every commit, and that definitely would have helped with some of the issues I faced. However, one of the most confusing parts of this was configuring the CLI to correctly talk with the NervesHub instance (NervesCloud in this case), and then realizing that I need to run `mix nerves_hub.user auth` (previously there were no error messages, just a crash that the `%NervesHubCLI.API.Auth{}` struct was `nil`). I did get it working eventually, but while talking to Josh and Lars, I realized that NervesHubCLI may have a better user experience if it was a globally-installable CLI tool, with some persistent settings.

## Potential solutions

There are a number of different ways which this problem could be tackled. 

- A hex archive - similar to the `phx_new` generators, or `nerves_bootstrap`. Installation could be as simple as `mix archive.install hex nerves_hub_cli`
- `#! /usr/bin/env elixir` script, with some `Mix.install` commands. 
- Modifying the project to support installation as an escript

For the first option of distributing it as a hex archive, I think this makes the most sense, and would be the easiest. However, hex archives are supposed to be lightweight extensions of `Mix`, so unfortunately, it can't have dependencies to prevent conflicts with the project deps. Given that the CLI has web requests and cryptography functions, I don't think it's a good idea to try to go without any dependencies

I made some preliminary experiments with the Elixir script idea, but I found that the `Mix` module was not available in the environment, which meant that I couldn't easily reuse the current mix tasks found in this repo, as they make heavy use of that module.

I think escript makes the most sense, as it can be easily distributed and doesn't require a large number of changes.

## Changes

To build the dependency as an escript, I basically copied all of the mix tasks into the `NervesHubCLI.CLI` module. I removed references to `Mix`, and swapped out the task-specific modules used (such as `lib/mix/nerves_hub/*.ex`) into variants which didn't have `Mix` dependencies. For example, instead of using `Mix.Shell.yes?/1`, I created a variant which just reads stdio via `IO.gets/1` and checks to see if it is a positive response.

The command structure is now slightly different. Because it is no longer invoked with mix tasks, the command follows a command + subcommand structure. Examples:

-   `$ nerves_hub user whoami`
-   `$ nerves_hub device list`
-   `$ nerves_hub key create --name dev_key`

Additionally, there are some persistent config options which can now be set. I anticipate support for these can be expanded in the future, but the main one is the `uri` option:

-   `$ nerves_hub config set uri "https://manage.nervescloud.com"`
-   `$ nerves_hub config get uri`

## What's broken

This PR is definitely not ready. I wanted to showcase this early work, to see if this CLI experience is something which would be a beneficial change for the community. As such, I spent time to get some basic functions to work, but didn't spend a lot of time to polish. Some things which are definitely broken:

-   The `:ca_store` dependency stores the cert files in the `priv/` directory - which is not available for escripts. So I commented out portions of the HTTP API interface which used that.
-   There are still some places where `Mix` is used. For example, try searching for `Mix.Project.config()` in the `lib/nerves_hub_cli/cli/` folder. If you run a command which calls `Mix` it will crash
-   I didn't update the error messages/help docs. Error messages will still refer to the mix tasks
-   Testing is non existent at this point (though 99% of the changes are coping code from the mix tasks into their own modules)
-   You may need to specify the flags for `org` or `product` (or set them as environment variables: `NERVES_HUB_ORG` and `NERVES_HUB_PRODUCT`), since they aren't read from application config. 

## How to test it out

I'm assuming you are using the NervesCloud instance here, if not change the uri to point to your own instance.

```
mix escript.install git https://github.com/gworkman/nerves_hub_cli branch gw-make-escript

# Note: you may need to add the escript directory to your path. Follow the instructions from escript

nerves_hub config set uri "https://manage.nervescloud.com/"
nerves_hub user whoami
nerves_hub user auth
nerves_hub user whoami
nerves_hub firmware list --org my_org --product my_product

# to uninstall
mix escript.uninstall nerves_hub
```

## What's next

If this is something which is useful for the community, I'll keep working on this to polish it up. For the most part, I haven't touched anything which breaks the mix tasks (just the removed `:ca_store`, and added global config of the `uri`).

I'm happy to get some feedback, and also some input on some of my opinionated things (such as naming the executable to `nerves_hub` or the module names - `NervesHubCLI.CLI.*` isn't great imo, but I wasn't sure where else to put it).

### CODE REVIEW

File: lib/nerves_hub_cli/api.ex [hunk 1]
```diff
@@ -157,11 +157,12 @@ defmodule NervesHubCLI.API do
       is_atom(ca_store) and !is_nil(ca_store) ->
         ca_store.ca_certs()
 
-      Code.ensure_loaded?(CAStore) ->
-        CAStore.file_path()
-        |> File.read!()
-        |> X509.from_pem()
-        |> Enum.map(&X509.Certificate.to_der/1)
+      # CAStore uses the priv directory, which does not work with escripts
```
Comment (joshk) [inline lib/nerves_hub_cli/api.ex:160]: this is an interesting issue.

**Comment by joshk** [comment]
I love where this is going!

Regarding setting the org and product, do you have any options/ideas to reduce the need to specify these on every cli use?

**Comment by gworkman** [comment]
Hey @joshk, thanks for the support! This is a really good question, but I have to start with the disclaimer that I am still quite new to NervesHub, so I haven't spent much time working with multiple orgs/products. But I did spend some time thinking about this use case, and here are several options for how I think we could support these settings:

1. For the casual user, or small teams who might only have one org and multiple products, or even just one product: allow configuration of organization and/or product at the global level, using `nerves_hub config set org my_org` and `nerves_hub config set product my_product`. But in this case, it would be very easy for users to inadvertently use the wrong configuration when starting new projects. So the CLI should probably notify the user of the global setting each time and maybe even ask for confirmation
2. For users who are switching between orgs and products frequently, I suggest that we write the docs to specifically mention the environment variables. That way it should be easy enough to `source .env` files, or use `direnv` or other tools to manage the project environment settings. This is actually already supported, so no changes are necessary
3. Another option would be to support a `--local` flag for the `nerves_hub config set org my_org` command. This could work in two ways: add a `nerves-hub.config` file in the current project directory which is checked when any commands are run, or, we could utilize the global configuration file to store the path to the current directory as a key with the config options to apply for the local directory (which is currently an Elixir map which gets stored and loaded with `:erlang.term_to_binary`). For example, the global config would look like:
```elixir
%{
  token: "nh_token_here",
  uri: "https://manage.nervescloud.com",
  local_config: %{
    "/path/to/my/project" => %{
      org: "my_org",
      product: "my_product"
    },
    "/path/to/other/project" => %{
      org: "my_other_org",
      product: "my_other_product"
    }
}
```

However, this solution breaks down when the project is moved, renamed, etc.

**Comment by joshk** [comment]
@gworkman I really love this! 

I'd be happy to merge this in ASAP and for us to continue to improve it.

I think we need to address the `CAStore` code that is commented out. Could we wrap it in an `if` and only load the certs from `CAStore` if the CLI isn't being used via escript?

File: lib/nerves_hub_cli/api.ex (hunk 1 above)
Comment (fhunleth) [reply lib/nerves_hub_cli/api.ex:160]: It looks like the OS-supplied certs are returned in the `true` clause. It seems like this code could be completely deleted.

Comment (joshk) [reply lib/nerves_hub_cli/api.ex:160]: you are right, i guess we can just state that you need to keep your os supplied certs up to date

File: lib/nerves_hub_cli/api.ex [hunk 2]
```diff
@@ -157,11 +162,10 @@ defmodule NervesHubCLI.API do
       is_atom(ca_store) and !is_nil(ca_store) ->
         ca_store.ca_certs()
 
-      Code.ensure_loaded?(CAStore) ->
-        CAStore.file_path()
-        |> File.read!()
-        |> X509.from_pem()
-        |> Enum.map(&X509.Certificate.to_der/1)
+      # TODO: These conditions can probably be removed after successfully testing for some time
```
Comment (gworkman) [inline lib/nerves_hub_cli/api.ex:165]: TODO: check security considerations of loading the CA certs at compile time into the binary, and the checks here

**Comment by joshk** [comment]
@gworkman can you also remove Elixir 1.11, 1.12, and 1.13 from the GitHub Actions configs?

File: lib/nerves_hub_cli/api.ex [hunk 3]
```diff
@@ -5,6 +5,11 @@ defmodule NervesHubCLI.API do
   @file_chunk 4096
   @progress_steps 50
 
+  @castore_certs CAStore.file_path()
+                 |> File.read!()
+                 |> X509.from_pem()
+                 |> Enum.map(&X509.Certificate.to_der/1)
+
```
Comment (joshk) [inline lib/nerves_hub_cli/api.ex:12]: @fhunleth @jjcarstens do either of you have any concerns with this?

Comment (fhunleth) [reply lib/nerves_hub_cli/api.ex:12]: It's ok. This is a really big PR that looks like a clear improvement so I mostly want to get it in and iterate.

**Comment by gworkman** [comment]
@joshk will do! I think it's about ready to merge in. There's one or two more spots I need to highlight for review

**Comment by gworkman** [comment]
@joshk updated min elixir version and CI config for 1.13, in line with what Frank said on slack that the rest of Nerves is >= 1.13

**Comment by lawik** [comment]
`help` doesn't seem to work. And all variants of `help <cmd>` or adding `-h`, `--help` in the command should ideally work. But we can do that next. This is so much smoother. Just tried it.

**Comment by lawik** [comment]
For the iterations:

If we could set config like `uri` as config per org I think lots of people get used to doing `-o community-fleet` or whatever as they work with Fly's tooling that generally requires that. We could try to do what they do and when not specified try to infer the config from the current project directory. But that might get messy.

**Review by gworkman — COMMENTED** [review]
Ready for review & merge. Please see the two highlighted points - 
`nerves_hub device burn`, `nerves_hub firmware sign` and `nerves_hub firmware publish` will all fail due to references to `Mix` module

**Comment by gworkman** [comment]
> `help` doesn't seem to work. And all variants of `help <cmd>` or adding `-h`, `--help` in the command should ideally work. But we can do that next. This is so much smoother. Just tried it.

Correct. Mix tasks had special docs so you could run `mix help TASK_NAME`, I didn't get around to implementing the equivalent feature yet. But it can be the next PR :)

**Comment by lawik** [comment]
> > `help` doesn't seem to work. And all variants of `help <cmd>` or adding `-h`, `--help` in the command should ideally work. But we can do that next. This is so much smoother. Just tried it.
> 
> Correct. Mix tasks had special docs so you could run `mix help TASK_NAME`, I didn't get around to implementing the equivalent feature yet. But it can be the next PR :)

Can definitely fix that later.

I did some poking at using Owl with this. Can rebase and merge that after this gets merged but for anyone curious:

https://github.com/lawik/nerves_hub_cli/tree/with-owl

I essentially implemented feature-parity and made the progress-bar a bit fancy :D

File: lib/nerves_hub_cli/cli/utils.ex [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    Shell.info("NervesHub organization: #{org}")
+    org
+  end
+
+  @doc """
+  Return the path to the generated firmware bundle
+  """
+  @spec firmware() :: Path.t()
+  def firmware do
+    # TODO: what to replace these paths with?
```
[the remark this answers is not in the document]
Comment (joshk) [reply lib/nerves_hub_cli/cli/utils.ex:93]: @lawik thoughts?

File: lib/nerves_hub_cli/cli/device.ex [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+    System.put_env("NERVES_HUB_CERT", File.read!(cert_path))
+    System.put_env("NERVES_HUB_KEY", File.read!(key_path))
+
+    burn_args =
+      opts
+      |> Keyword.take([:firmware])
+      |> OptionParser.to_argv()
+
+    # TODO: remove reference to mix task
+    Mix.Task.run("burn", burn_args)
```
[the remark this answers is not in the document]
Comment (joshk) [reply lib/nerves_hub_cli/cli/device.ex:406]: @lawik thoughts?
I think using `System.cmd` shouldn't be fine, although I'd need to defer to @fhunleth , @jjcarstens and @lawik

[the remark this answers is not in the document]
Comment (jjcarstens) [reply lib/nerves_hub_cli/cli/device.ex:406]: The modules should be defined right? Would you be able to call `Mix.Tasks.Burn.run(burn_args)` directly?

[the remark this answers is not in the document]
Comment (gworkman) [reply lib/nerves_hub_cli/cli/device.ex:406]: @jjcarstens The main change here is converting everything to an escript, which is compiled into a binary and doesn't have access to the `Mix` module. So unfortunately that doesn't work.

If you run a command which tries to call a `Mix` function, you get the error `module Mix.Project is not available` or `module Mix.Tasks.Burn` is not available.

**Comment by joshk** [comment]
The only thing I think is missing before I hit merge is an updated readme that explains the recommended installation steps.

**Comment by gworkman** [comment]
@joshk @lawik Ready to merge :)

- updated README with instructions for escript installation, also added instructions for how to configure the NervesHub URI and authenticate
- replaced `Mix.Task.run` with `System.cmd` for the `nerves_hub device burn FIRMWARE` command
- removed the no-argument commands for `nerves_hub firmware sign` and `nerves_hub firmware publish`. Now the path to the firmware is always required, because we can't infer it from the build path (previously obtained via `Mix.Project.build_path()`)

**Comment by joshk** [comment]
This is awesome!

I'm happy to merge this in as is. 

Before I do, I did want to check one thing: for `nerves_hub firmware publish`, is there any way we could infer a default path to use if for the 'no-argument' command?

**Comment by gworkman** [comment]
@joshk I think it might be a good issue to leave for a future version.

I'd recommend users specify the firmware build path. mix firmware accepts a path argument:

--output - (Optional) The path to the .fw file used to write the patch firmware. Defaults to Nerves.Env.firmware_path/1

(I can put this in the docs somewhere)

If we want to detect the path, we have to worry about reading the mix.exs file and rebuilding the config from the environment. If we take a more naive approach, then we need to make assumptions about the MIX_ENV, build paths, etc. 

Maybe the amount of effort to read the mix.exs file and extract the config is easier than I expect, and would love to learn more about that process if that's the case 😃 but for now I think let's get this out there and have a further discussion on how to extract the firmware path

**Review by joshk — APPROVED** [review]
Let's do it!
